Microsoft.Toolkit.HighPerformance/Helpers/ParallelHelper.ForEach.IInAction2D.cs

```diff
@@ -0,0 +1,158 @@
+        /// <param name="action">The <typeparamref name="TAction"/> instance representing the action to invoke.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ForEach<TItem, TAction>(ReadOnlyMemory2D<TItem> memory, in TAction action)
+            where TAction : struct, IInAction<TItem>
+        {
+            ForEach(memory, action, 1);
+        }
+
+        /// <summary>
+        /// Executes a specified action in an optimized parallel loop over the input data.
+        /// </summary>
+        /// <typeparam name="TItem">The type of items to iterate over.</typeparam>
+        /// <typeparam name="TAction">The type of action (implementing <see cref="IInAction{T}"/> of <typeparamref name="TItem"/>) to invoke over each item.</typeparam>
+        /// <param name="memory">The input <see cref="ReadOnlyMemory2D{T}"/> representing the data to process.</param>
+        /// <param name="action">The <typeparamref name="TAction"/> instance representing the action to invoke.</param>
+        /// <param name="minimumActionsPerThread">
+        /// The minimum number of actions to run per individual thread. Set to 1 if all invocations
+        /// should be parallelized, or to a greater number if each individual invocation is fast
+        /// enough that it is more efficient to set a lower bound per each running thread.
+        /// </param>
+        public static void ForEach<TItem, TAction>(ReadOnlyMemory2D<TItem> memory, in TAction action, int minimumActionsPerThread)
+            where TAction : struct, IInAction<TItem>
+        {
+            if (minimumActionsPerThread <= 0)
+            {
+                ThrowArgumentOutOfRangeExceptionForInvalidMinimumActionsPerThread();
+            }
+
+            if (memory.IsEmpty)
+            {
+                return;
+            }
+
+            int
+                maxBatches = 1 + ((memory.Size - 1) / minimumActionsPerThread),
+                clipBatches = Math.Min(maxBatches, memory.Height),
+                cores = Environment.ProcessorCount,
+                numBatches = Math.Min(clipBatches, cores),
+                batchHeight = 1 + ((memory.Height - 1) / numBatches);
+
+            var actionInvoker = new InActionInvokerWithReaadOnlyMemory2D<TItem, TAction>(batchHeight, memory, action);
+
+            // Skip the parallel invocation when possible
+            if (numBatches == 1)
+            {
+                actionInvoker.Invoke(0);
+
+                return;
+            }
+
+            // Run the batched operations in parallel
+            Parallel.For(
+                0,
+                numBatches,
+                new ParallelOptions { MaxDegreeOfParallelism = numBatches },
+                actionInvoker.Invoke);
+        }
+
+        // Wrapping struct acting as explicit closure to execute the processing batches
+        private readonly struct InActionInvokerWithReaadOnlyMemory2D<TItem, TAction>
```

Comment:
This type should not be readonly as its members shouldn't be

Reply:
Discussed in chat in the UWP Community Discord server, resolving these two comments.